### PR TITLE
doc: Add Egress Gateway Getting Started Guide

### PR DESCRIPTION
**Pull request by blzhao-0**
This is a documentation of the detailed instructions on configuring
cilium components to start using the egress nat gateway feature.

RFE: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>
Signed-off-by: Bolun Zhao <blzhao@google.com>

### CODE REVIEW

File: Documentation/gettingstarted/egress-gateway.rst [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+.. _egress_gateway:
+
+**********************
+Egress Gateway (beta)
+**********************
+
+Egress gateway provides a way to configure the egress policy. Without
+this feature, pod traffic could only egress from the node it is running
+on and masqueraded with the node IP.
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
on and would be masqueraded with the node IP.
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+.. include:: ../beta.rst
+
+Configure Cilium
+================
+
+.. include:: k8s-install-download-release.rst
+
+The Cilium Egress Gateway feature relies on :ref:`kubeproxy-free`,
+follow the guide to create a new deployment. The beta feature is disabled by default.
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
follow the guide to create a new deployment. The feature is disabled by default.
```
No need to remind that it is in beta mode here. Plus this is likely something we would forget to remove when moving out of beta.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+Enable the feature by setting the ``localRedirectPolicy`` value to ``true``.
+
+.. parsed-literal::
+
+   helm install cilium |CHART_RELEASE| \\
+     --set enableEgressGateway=true
+
+Verify that Cilium agent pod is running.
+
+.. code-block:: bash
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
.. code-block:: shell-session
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    $ kubectl -n kube-system get pods -l k8s-app=cilium
+    NAME           READY   STATUS    RESTARTS   AGE
+    cilium-5ngzd   1/1     Running   0          3m19s
+
+Validate that the Cilium Egress NAT Policy CRD has been registered.
+
+.. code-block:: bash
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
.. code-block:: shell-session
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	   [...]
+	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z
+
+
+Optionally, Egress Gateway support can be enabled by setting
+``enable-egress-gateway`` to ``true`` in ConfigMap. By default, it is disabled.
+
+To set ``enable-egress-gateway`` to ``true``, it requires
+``enable-bpf-masqurade`` and ``kube-proxy-replacement``. It cannot be enabled
+with these 2 options set as ``false`` or ``disabled``.
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
with either of these 2 options set as ``false`` or ``disabled``.
```
What about KPR in partial mode?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+Architecture
+============
+
+Overall, some nodes in the cluster will be assigned as gateway nodes. These gateway nodes not
+only have node IP, but also have additional pools of egress IPs attached to it. These egress IP
+addresses can be associated with certain workloads (pods) that run in the cluster. When certain
+packets are selected, instead of exitting the cluster immediatly, they will be forwarded to gateway
+node and SNAT with egress IP there. A tunnel interface will be created when Egress Gateway is
+enabled. The tunnel is for the forwarding in Egress Gateway. All other traffic will keep their
+own behavior based on which tunnel mode is selected.
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
Overall, some nodes in the cluster will be assigned as gateway nodes. These gateway nodes not
only have a node IP, but also have additional pools of egress IPs attached to it. These egress IP
addresses can be associated with certain workloads (pods) that run in the cluster. When certain
packets are selected, instead of exiting the cluster immediately, they will be forwarded to a gateway
node and SNAT-ed with an egress IP there. A tunnel interface will be created when Egress Gateway is
enabled. The tunnel is for forwarding with Egress Gateway. All other traffic will keep their
own behavior based on which tunnel mode is selected.
```

`some nodes in the cluster` as well as `When certain packets are selected` are not precise. What nodes? What packets?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+.. image:: images/cilium_egress_gateway_flow.png
+   :scale: 70%
+
+.. image:: images/cilium_egress_gateway_return_flow.png
+   :scale: 70%
+
+
+
+Create external service
+=======================
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
Create an External Service
==========================
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    ● nginx.service - A high performance web server and a reverse proxy server
+    Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
+    Active: active (running) since Sun 2021-04-04 21:58:57 UTC; 1min 3s ago
+    [...]
+    $ curl http://192.168.33.13:80  # Assume 192.168.33.13 is the external IP of the node
+    [...]
+    <title>Welcome to nginx!</title>
+    [...]
+
+Create client pods
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
Create Client Pods
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+Create an nginx service on a Linux node that is external to the existing Kubernetes
+cluster, and use it as the destination of the egress traffic.
+
+.. code-block:: shell-session
+
+    $ # Install and start nginx
+    $ sudo apt install nginx
+    $ sudo systemctl start nginx
+
+    $ # Make sure the service is started and listen on port :80
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
    $ # Make sure the service is started and listens on port :80
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    $ curl http://192.168.33.13:80  # Assume 192.168.33.13 is the external IP of the node
+    [...]
+    <title>Welcome to nginx!</title>
+    [...]
+
+Create client pods
+==================
+
+Deploy a client pod from which egress traffic is generated, and try to access previous
+external service
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
external service.
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+	   NAME                              CREATED AT
+	   [...]
+	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z
+
+
+Optionally, Egress Gateway support can be enabled by setting
+``enable-egress-gateway`` to ``true`` in ConfigMap. By default, it is disabled.
+
+To set ``enable-egress-gateway`` to ``true``, it requires
+``enable-bpf-masqurade`` and ``kube-proxy-replacement``. It cannot be enabled
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
``enable-bpf-masquerade`` and ``kube-proxy-replacement``. It cannot be enabled
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+.. literalinclude:: ../../examples/kubernetes-egress-gateway/egress-ip-deployment.yaml
+
+
+Create Egress NAT Policy
+========================
+
+Apply the following Egress NAT Policy, which basically means: when the pod is running in the namespace
+with label ``ns=default`` and the pod itself has label ``run=client``, if it's trying to talk to IP CIDR
+``192.168.33.13/32``, use egress IP ``192.168.33.100``. In this example, it tells cilium to forward the
+packet from pod ``client`` to the gateway node with egress IP ``192.168.33.100``, and masqurade with that
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
packet from pod ``client`` to the gateway node with egress IP ``192.168.33.100``, and masquerade with that
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+.. literalinclude:: ../../examples/kubernetes-egress-gateway/egress-ip-deployment.yaml
+
+
+Create Egress NAT Policy
+========================
+
+Apply the following Egress NAT Policy, which basically means: when the pod is running in the namespace
+with label ``ns=default`` and the pod itself has label ``run=client``, if it's trying to talk to IP CIDR
+``192.168.33.13/32``, use egress IP ``192.168.33.100``. In this example, it tells cilium to forward the
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
``192.168.33.13/32``, use egress IP ``192.168.33.100``. In this example, it tells Cilium to forward the
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+.. code-block:: shell-session
+
+    $ kubectl exec -it client --image=busybox
+    If you don't see a command prompt, try pressing enter.
+    / # wget http://192.168.33.13:80 -o /dev/null
+
+Verify access log from nginx node that the request is coming from egress IP now instead of one of the
+nodes in Kubernetes cluster.
+
+    .. code-block:: shell-session
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]: Wrong indent level.

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+          value: "192.168.33.100/24 192.168.33.101/24"
+        args:
+        - "for i in $EGRESS_IPS; do ip address add $i dev enp0s8; done; sleep 10000000"
+        lifecycle:
+          preStop:
+            exec:
+              command:
+              - "/bin/sh"
+              - "-c"
+              - "for i in $EGRESS_IPS; do ip address del $i dev enp0s8; done"
```
Comment (qmonnet) [inline examples/kubernetes-egress-gateway/egress-ip-deployment.yaml]: Nit: missing end-of-line character at the end of this file.
(Same applies to egress-nat-policy.yaml.)

File: Documentation/gettingstarted/egress-gateway.rst [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+        $ tail /var/log/nginx/access.log
+        [...]
+        192.168.33.100 - - [04/Apr/2021:22:06:57 +0000] "GET / HTTP/1.1" 200 612 "-" "Wget/1.19.4 (linux-gnu)"
+
+
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks good overall, I find the steps very clear and well-explained. I have a few comments on the formatting below.

Please also address the issues reporting from the CI:
- A few typo (also highlighted in my review)
- You need to reference your documents from an index. From the list of the Getting Started Guides for example, likely in the [Advanced Networking](https://docs.cilium.io/en/latest/gettingstarted/#advanced-networking) section.

Please also make sure that you use consistent levels of indents for all the directives in your document (block-content should ideally be left-aligned on the first letter of the directive name, as is the case for your first parsed-literal block for example).

File: Documentation/gettingstarted/egress-gateway.rst (hunk 16 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: Fixed. Thanks!

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml (hunk 15 above)
Comment (blzhao-0) [reply examples/kubernetes-egress-gateway/egress-ip-deployment.yaml]: Added newline.
Thx.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 14 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: This is fixed.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 5 above)
Comment (qmonnet) [reply Documentation/gettingstarted/egress-gateway.rst]: What feature does it need exactly? Can kube-proxy replacement be used in partial mode, or does it have to be strict mode?

File: Documentation/gettingstarted/egress-gateway.rst (hunk 6 above)
Comment (qmonnet) [reply Documentation/gettingstarted/egress-gateway.rst]: Can you please address the observation above?

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks for the update! There are still a few items remaining, please see the un-resolved items above. In addition, you still need to link the document somewhere, and to add `busybox` to the list of correct words as suggested by the failing GitHub action:

```
 Please fix the following documentation warnings:
/github/workspace/Documentation/gettingstarted/egress-gateway.rst: WARNING: document isn't included in any toctree

Please fix the following spelling mistakes:
* Documentation/gettingstarted/egress-gateway.rst:127: (busybox) 

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh busybox 
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+    cilium-5ngzd   1/1     Running   0          3m19s
+
+Validate that the Cilium Egress NAT Policy CRD has been registered.
+
+.. code-block:: shell-session
+
+	   $ kubectl get crds
+	   NAME                              CREATED AT
+	   [...]
+	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z
```
Comment (qmonnet) [inline Documentation/gettingstarted/egress-gateway.rst]: Additionally, please use consistent indentation levels throughout the document. This one in particular stands out with the rest, but most of your code blocks have 4 spaces when they should have 3, like for your `parsed-literal` block.

Comment (joestringer) [reply Documentation/gettingstarted/egress-gateway.rst]: +1, sphinx is quite picky about the indentation unfortunately.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+.. _egress_gateway:
+
+**********************
+Egress Gateway (beta)
+**********************
+
+Egress gateway provides a way to configure the egress policy. Without
+this feature, pod traffic could only egress from the node it is running
+on and would be masqueraded with the node IP.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: I would turn this other way around:
```
The egress gateway allows users to redirect egress pod traffic through
specific, gateway nodes. Packets are masqueraded to the gateway node IP.
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+This document explains how to configure Cilium to enable egress gateway
+as well as how to configure egress NAT policy to route and SNAT the egress
+traffic for a specific workload.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
This document explains how to enable the egress gateway and configure
egress NAT policies to route and SNAT the egress traffic for a specific
workload.
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+.. include:: ../beta.rst
+
+Configure Cilium
+================
+
+.. include:: k8s-install-download-release.rst
+
+The Cilium Egress Gateway feature relies on :ref:`kubeproxy-free`,
+follow the guide to create a new deployment. The feature is disabled by default.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: The kube-proxy-free guide instructs users to remove kube-proxy completely. I'm worried we might give the wrong impression here, that kube-proxy needs to be removed. Maybe we can just say it require our kube-proxy replacement (with a link) and add `kubeProxyReplacement=strict` in the command below.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+Verify that Cilium agent pod is running.
+
+.. code-block:: shell-session
+
+    $ kubectl -n kube-system get pods -l k8s-app=cilium
+    NAME           READY   STATUS    RESTARTS   AGE
+    cilium-5ngzd   1/1     Running   0          3m19s
+
+Validate that the Cilium Egress NAT Policy CRD has been registered.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: I would drop this or, better, move it to a troubleshooting section. For a getting started guide, I think we want to take the shortest route. If the CRD is missing it's a bug and therefore corner case.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+.. include:: ../beta.rst
+
+Configure Cilium
+================
+
+.. include:: k8s-install-download-release.rst
+
+The Cilium Egress Gateway feature relies on :ref:`kubeproxy-free`,
+follow the guide to create a new deployment. The feature is disabled by default.
+Enable the feature by setting the ``localRedirectPolicy`` value to ``true``.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]:
```suggestion
Enable the feature by setting the ``enableEgressGateway`` value to ``true``.
```

File: Documentation/gettingstarted/egress-gateway.rst [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+
+.. parsed-literal::
+
+   helm install cilium |CHART_RELEASE| \\
+     --set enableEgressGateway=true
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: This Helm option is currently not in master. We need the docs PR for Cilium v1.10-rc1.
@anfernee Could you submit the Helm option in a separate PR? I think it's already in the e2e test PR, right?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+	   [...]
+	   ciliumegressnatpolicy.cilium.io   2021-03-26T02:11:21Z
+
+
+Optionally, Egress Gateway support can be enabled by setting
+``enable-egress-gateway`` to ``true`` in ConfigMap. By default, it is disabled.
+
+To set ``enable-egress-gateway`` to ``true``, it requires
+``enable-bpf-masquerade`` and ``kube-proxy-replacement``. It cannot be enabled
+with either of these 2 options set as ``false`` or ``disabled``.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: I would drop this paragraph and just add a note on line 28 to say it also requires BPF masquerading.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+own behavior based on which tunnel mode is selected.
+
+The following 2 diagrams show the packet flow for both egress and return traffic when using direct
+routing mode. Note that the return packet is not going over a tunnel, while forwarding traffic is. 
+
+.. image:: images/cilium_egress_gateway_flow.png
+   :scale: 70%
+
+.. image:: images/cilium_egress_gateway_return_flow.png
+   :scale: 70%
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: This section is great! But it's usually something we include in the blog post for the new release (I'll ping you when we start writing that). If we include it in the documentation, then (1) it should probably be somewhere in Concepts instead and (2) that means we'll try to maintain it. I would just remove this and keep it on the side for the blog post.

File: Documentation/gettingstarted/egress-gateway.rst [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+
+
+
+Create an External Service
+==========================
+
+Create an nginx service on a Linux node that is external to the existing Kubernetes
+cluster, and use it as the destination of the egress traffic.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst:56]: I understand this is useful to show that the received packets have been masqueraded to the expected egressIP, but most users testing this guide are likely to do so on managed Kubernetes platforms where reproducing this is not as straightforward.

Do we really need to show the source IP in the nginx logs? Could we instead show in some other way that the packets have gone through the egress gateway node as expected?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+==================
+
+Deploy a client pod from which egress traffic is generated, and try to access previous
+external service.
+
+.. code-block:: shell-session
+
+    $ kubectl run -it client --image=busybox
+    If you don't see a command prompt, try pressing enter.
+    / # wget http://192.168.33.13:80 -o /dev/null
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: AFAIU, this will require users to switch between terminals to follow the guide. Could we instead use a proper Pod manifest? There are multiple examples in other guides (LRP guide).

EDIT: I see you're already using a manifest later on. Maybe we could use it sooner?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+
+Verify access log from previous node that the request is coming from one of the node in
+Kubernetes cluster.
+
+.. code-block:: shell-session
+
+    $ tail /var/log/nginx/access.log
+    192.168.33.11 - - [04/Apr/2021:22:06:57 +0000] "GET / HTTP/1.1" 200 612 "-" "Wget/1.19.4 (linux-gnu)"
+
+In the previous example, the busybox is running on the node ``192.168.33.11``, so the result makes sense.
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: I think we can skip this first check. This is essentially testing Cilium works, which is out of scope for the guide IMO. If you really want to show before/after, I would configure everything, show before, add the label, show after.

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+      - name: egress-ip
+        image: docker.io/library/busybox:1.31.1
+        command: ["/bin/sh","-c"]
+        securityContext:
+          privileged: true
+        env:
+        - name: EGRESS_IPS
+          value: "192.168.33.100/24 192.168.33.101/24"
+        args:
+        - "for i in $EGRESS_IPS; do ip address add $i dev enp0s8; done; sleep 10000000"
```
Comment (pchaigno) [inline examples/kubernetes-egress-gateway/egress-ip-deployment.yaml:28]: Is this going to be enough for users running in e.g. GKE, EKS, AKS?

**Review by pchaigno — CHANGES_REQUESTED** [review]
I fixed the couple build issues mentioned by Quentin to be able to render locally and review. I haven't tested yet, but I think it needs a couple changes if we want folks to be able to follow the guide (without needing to deploy our development VM).

File: Documentation/gettingstarted/egress-gateway.rst (hunk 5 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: This feature is designed and tested under kube-proxy-replacement strict mode. I don't know if we can support its functionality when kube-proxy is partially present. @anfernee could you comment on this?

Comment (pchaigno) [reply Documentation/gettingstarted/egress-gateway.rst]: `=partial` won't work with the egress gateway unless all KPR sub-features are enabled, so I think we should just recommend to set `=strict`.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 18 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: agreed, fixed.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 19 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: fixed.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 22 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: fixed.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 21 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: removed for now.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 24 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: done.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 26 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst:56]: I think this is just for an example of verification that is easy to understand and easy to use in practice.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 25 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: I think it makes sense. Removed for this guide.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 17 above)
Comment (blzhao-0) [reply Documentation/gettingstarted/egress-gateway.rst]: fixed the consistency issue. thx!

File: Documentation/gettingstarted/egress-gateway.rst (hunk 26 above)
Comment (pchaigno) [reply Documentation/gettingstarted/egress-gateway.rst:56]: I agree it's a good example, but how are users with e.g. a GKE cluster supposed to follow this guide?

Comment (anfernee) [reply Documentation/gettingstarted/egress-gateway.rst:56]: They should be able to create a vm instance and install nginx. But you are right, this should be marked as optional. The user should be able to use their own external service to verify.

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml (hunk 29 above)
Comment (anfernee) [reply examples/kubernetes-egress-gateway/egress-ip-deployment.yaml:28]: This is the guide for onprem environment. In cloud, the way to assign IP addresses will be very different in different cloud providers. We haven't covered those environments yet. Maybe we could do it as next step.

File: Documentation/gettingstarted/egress-gateway.rst (hunk 26 above)
Comment (pchaigno) [reply Documentation/gettingstarted/egress-gateway.rst:56]: I'm fine if we mark that step as optional.

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml (hunk 29 above)
Comment (pchaigno) [reply examples/kubernetes-egress-gateway/egress-ip-deployment.yaml:28]: Maybe it would be enough to use the existing node IP for managed Kubernetes? It won't demonstrate the SNAT part of the egress gateway, but at least users would be able to observe the redirect. I guess it's better than nothing.

If that's not an option, we should at least clarify at the top that the guide doesn't currently support managed Kubernetes. I would be very careful how we word this though; we don't want to give the impression the feature doesn't support managed Kubernetes.

**Review by qmonnet — APPROVED** [review]
Looks good from my side. Thanks for the updates.

File: examples/kubernetes-egress-gateway/egress-ip-deployment.yaml (hunk 29 above)
Comment (anfernee) [reply examples/kubernetes-egress-gateway/egress-ip-deployment.yaml:28]: Make sense. I'll try in on GKE first as an example. But I won't be able to cover other platforms.

Comment (pchaigno) [reply examples/kubernetes-egress-gateway/egress-ip-deployment.yaml:28]: > But I won't be able to cover other platforms.

I think that's fine as long as we don't make the guide GKE-specific in some way. If there are small issues when running it on other managed K8s, we can always fix them up later.

File: examples/kubernetes-egress-gateway/egress-nat-policy.yaml [hunk 30]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  name: egress-sample
+spec:
+  egress:
+  - podSelector:
+      matchLabels:
+        class: mediabot
+    namespaceSelector:
+      matchLabels:
+        ns: default
```
Comment (jrajahalme) [inline examples/kubernetes-egress-gateway/egress-nat-policy.yaml]: Wouldn't it be simpler to add namespace label match (`io.kubernetes.pod.namespace: default`) to the podSelector instead of using a namespaceSelector? Or is the CiliumEgressNATPolicy somehow applied on the namespace directly, rather than just on the selected pods?

**Review by jrajahalme — APPROVED** [review]
One question for potential simplification only.

File: examples/kubernetes-egress-gateway/egress-nat-policy.yaml (hunk 30 above)
Comment (anfernee) [reply examples/kubernetes-egress-gateway/egress-nat-policy.yaml]: namespace label is the field that has direct support from api, which we showed in the guide. `io.kubernetes.pod.namespace: default` probably will also work. @MasterZ40 keep me honest here.

Comment (blzhao-0) [reply examples/kubernetes-egress-gateway/egress-nat-policy.yaml]: So as previously written the watcher is indeed directly invoked by namespace label updates, however I have changed the approach to watch directly on CEP objects for more scalability. Therefore i think @jrajahalme is correct here we should move ns name references to use pod selector labels and leave the namespace selector to be empty. I just tested that the "io.kubernetes.pod.namespace: default" approach works for this purpose.

We can also see there the CEP labels with namespace prefix here doesn't actually contain ns name.
```
k8s:class=mediabot
k8s:io.cilium.k8s.namespace.labels.bar=Foo
k8s:io.cilium.k8s.namespace.labels.foo=Bar
k8s:io.cilium.k8s.namespace.labels.user=Alice
k8s:io.cilium.k8s.policy.cluster=default
k8s:io.cilium.k8s.policy.serviceaccount=default
k8s:io.kubernetes.pod.namespace=default
k8s:org=empire
```

Comment (blzhao-0) [reply examples/kubernetes-egress-gateway/egress-nat-policy.yaml]: In essence, both approach works. And yes we should probably also explain the "io.kubernetes.pod.namespace: default" somewhere in the guide.

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
test-me-please

**Comment by pchaigno** [comment]
> test-me-please

@anfernee This isn't needed for documentation-only changes. There's already `Documentation Updates / Validate & Build HTML (pull_request)`, which runs automatically, to test such changes.

**Review by pchaigno — APPROVED** [review]
